qiskit_ode/signals/signals.py

```diff
@@ -11,183 +11,230 @@
+        return self.envelope(t) * np.exp(arg)
 
-    def __add__(self, other):
-        return signal_add(self, other)
+    def __call__(self, t: Union[float, np.array, Array]) -> Union[complex, np.array, Array]:
+        """Vectorized evaluation of the signal at time t."""
+        return np.real(self.complex_value(t))
 
-    def __radd__(self, other):
-        return signal_add(self, other)
+    def __str__(self) -> str:
+        """Return string representation."""
+        if self.name is not None:
+            return str(self.name)
 
-    def plot(self, t0: float, tf: float, n: int, axis=None):
-        """Plot the mixed signal over an interval.
+        return "Signal(carrier_freq={freq}, phase={phase})".format(
+            freq=str(self.carrier_freq), phase=str(self.phase)
+        )
 
-        Args:
-            t0: initial time
-            tf: final time
-            n: number of points to sample in interval.
-            axis (AxesSubplot): the axis to use for plotting.
-        """
-        x_vals = np.linspace(t0, tf, n)
+    def __add__(self, other: "Signal") -> "SignalSum":
+        return signal_add(self, other)
 
-        sig_vals = []
-        for x in x_vals:
-            sig_vals.append(self.value(x))
+    def __radd__(self, other: "Signal") -> "SignalSum":
+        return self.__add__(other)
 
-        if axis:
-            axis.plot(x_vals, sig_vals)
-        else:
-            plt.plot(x_vals, sig_vals)
+    def __mul__(self, other: "Signal") -> "SignalSum":
+        return signal_multiply(self, other)
 
-    def plot_envelope(self, t0: float, tf: float, n: int, axis=None):
-        """Plot the envelope over an interval.
+    def __rmul__(self, other: "Signal") -> "SignalSum":
+        return self.__mul__(other)
 
-        Args:
-            t0: initial time
-            tf: final time
-            n: number of points to sample in interval.
-            axis (AxesSubplot): the axis to use for plotting.
-        """
-        x_vals = np.linspace(t0, tf, n)
+    def __neg__(self) -> "SignalSum":
+        return -1 * self
 
-        sig_vals = []
-        for x in x_vals:
-            sig_vals.append(self.envelope_value(x))
+    def __sub__(self, other: "Signal") -> "SignalSum":
+        return self + (-other)
 
-        if axis:
-            axis.plot(x_vals, np.real(sig_vals))
-            axis.plot(x_vals, np.imag(sig_vals))
-        else:
-            plt.plot(x_vals, np.real(sig_vals))
-            plt.plot(x_vals, np.imag(sig_vals))
+    def __rsub__(self, other: "Signal") -> "SignalSum":
+        return other + (-self)
 
+    def conjugate(self):
+        """Return a new signal whose complex value is the complex conjugate of this one."""
 
-class Signal(BaseSignal):
-    """The most general mixed signal type, represented by a callable
-    envelope function and a carrier frequency.
-    """
+        def conj_env(t):
+            return np.conjugate(self.envelope(t))
```

Comment:
Yeah: in the `__init__` I've set it so that if jax is being used, the output of the envelope function will always be wrapped in an `Array`, so here the dispatch will implement `np.conjugate` via `jnp.conjugate`.

Reply:
Thanks.